File: docs/examples/chatbot-memory.py
"""
Chatbot Memory Example

This example demonstrates:
- Building a stateful chatbot with RecallBricks
- Storing conversation history
- Retrieving relevant context
- Using predictive recall for smart responses
- Session management
"""

from recallbricks import RecallBricks
import os
from datetime import datetime
import time

rb = RecallBricks(os.getenv('RECALLBRICKS_API_KEY'))

# Simulated user sessions
user_id = 'user_12345'
session_id = f'session_{int(time.time())}'

def main():
    try:
        print('💬 RecallBricks Chatbot Memory Example\n')
        print(f'User: {user_id}')
        print(f'Session: {session_id}\n')

        # ============================================
        # Initialize: Store user preferences
        # ============================================
        print('=== Session Start ===\n')

        rb.memories.create_batch([
            {
                'content': 'User prefers concise responses',
                'metadata': {
                    'user_id': user_id,
                    'type': 'preference',
                    'category': 'communication_style'
                }
            },
            {
                'content': 'User is a senior software engineer',
                'metadata': {
                    'user_id': user_id,
                    'type': 'profile',
                    'category': 'professional_info'
                }
            },
            {
                'content': 'User timezone: PST (UTC-8)',
                'metadata': {
                    'user_id': user_id,
                    'type': 'profile',
                    'category': 'location'
                }
            }
        ])

        print('✓ User preferences initialized\n')

        # ============================================
        # TURN 1: First message
        # ============================================
        print('--- Turn 1 ---\n')
        handle_message('What are the best practices for API design?', 1)
        print()

        # ============================================
        # TURN 2: Follow-up question
        # ============================================
        print('--- Turn 2 ---\n')
        handle_message('How about authentication specifically?', 2)
        print()

        # ============================================
        # TURN 3: Reference to earlier context
        # ============================================
        print('--- Turn 3 ---\n')
        handle_message('Can you give me a code example?', 3)
        print()

        # ============================================
        # Session End: Summarize conversation
        # ============================================
        print('=== Session End ===\n')

        conversation_memories = rb.memories.search(
            query='conversation',
            metadata={
                'user_id': user_id,
                'session_id': session_id
            },
            weights={'semantic': 0.3, 'recency': 0.7},
            limit=10
        )

        print('Conversation Summary:')
        print(f'  Total turns: 3')
        print(f'  Topics discussed: API design, authentication, code examples')
        print(f'  Memories stored: {len(conversation_memories)}\n')

        # ============================================
        # New Session: Returning user
        # ============================================
        print('=== New Session (User Returns) ===\n')

        new_session_id = f'session_{int(time.time()) + 1000}'

        print(f'New Session: {new_session_id}\n')
        print('User: "Hey, I\'m back!"\n')

        # Use predictive recall to understand what user might need
        prediction = rb.metacognition.predict(
            context=f"""
                Returning user greeting.
                User ID: {user_id}
                Previous session: Discussed API design and authentication
                Session start
            """
        )

        print('AI Prediction for returning user:')
        print(f'  Confidence: {prediction.confidence * 100:.1f}%\n')

        print('Suggested Context:')
        high_confidence_suggestions = [
            m for m in prediction.suggested_memories
            if m.confidence > 0.8
        ][:3]

        for index, mem in enumerate(high_confidence_suggestions):
            print(f'  {index + 1}. {mem.content}')
            print(f'     Confidence: {mem.confidence * 100:.1f}%')

        print('\nBot Response:')
        print('  "Welcome back! Last time we discussed API design and authentication.')
        print('   Would you like to continue with code examples?"')
        print()

        # ============================================
        # Summary
        # ============================================
        print('✅ Chatbot Memory Example Completed!\n')
        print('Key features demonstrated:')
        print('  • Store conversation history automatically')
        print('  • Retrieve relevant context using semantic search')
        print('  • Use predictive recall for smart suggestions')
        print('  • Maintain user preferences across sessions')
        print('  • Track conversation flow with metadata')
        print('  • Welcome returning users with context awareness')

    except Exception as error:
        print(f'❌ Error: {str(error)}')
        exit(1)

def handle_message(user_message, turn_number):
    """Handle a single chatbot turn"""
    print(f'User: "{user_message}"')

    # 1. Use predictive recall to get relevant context
    prediction = rb.metacognition.predict(
        context=f"""
            User message: "{user_message}"
            Turn: {turn_number}
            User ID: {user_id}
            Session ID: {session_id}
        """,
        limit=3,
        min_confidence=0.7
    )

    print('\nRelevant Context Retrieved:')
    for index, mem in enumerate(prediction.suggested_memories):
        print(f'  {index + 1}. {mem.content}')

    # 2. Generate response (simulated)
    bot_response = generate_response(user_message, prediction.suggested_memories)

    print(f'\nBot: "{bot_response}"')

    # 3. Store this exchange as memory
    rb.memories.create_batch([
        {
            'content': f'User asked: "{user_message}"',
            'metadata': {
                'user_id': user_id,
                'session_id': session_id,
                'turn': turn_number,
                'type': 'user_message',
                'timestamp': datetime.utcnow().isoformat()
            }
        },
        {
            'content': f'Bot responded: "{bot_response}"',
            'metadata': {
                'user_id': user_id,
                'session_id': session_id,
                'turn': turn_number,
                'type': 'bot_message',
                'timestamp': datetime.utcnow().isoformat()
            }
        }
    ])

    print('\n✓ Conversation stored in memory')

def generate_response(message, context):
    """Simulate response generation"""
    # In a real chatbot, this would use an LLM with the retrieved context
    message_lower = message.lower()

    if 'api design' in message_lower:
        return 'Best practices for API design include: RESTful principles, versioning, clear documentation, consistent naming, and proper error handling.'
    elif 'authentication' in message_lower:
        return 'For authentication, I recommend using OAuth 2.0 or JWT tokens. Bearer tokens in the Authorization header are standard practice.'
    elif 'code example' in message_lower:
        return 'Here\'s a quick example: Authorization: Bearer your_token_here'
    else:
        return 'I can help you with that! Let me provide some information based on your preferences for concise responses.'

if __name__ == '__main__':
    main()

"""
To run this example:

1. Install dependencies:
   pip install recallbricks python-dotenv

2. Set your API key:
   export RECALLBRICKS_API_KEY='rb_live_your_key_here'

3. Run the script:
   python chatbot-memory.py

This demonstrates:
  - How to build a stateful chatbot
  - Automatic context retrieval
  - Cross-session memory
  - Personalized responses based on preferences
  - Smart predictions for returning users
"""

File: docs/examples/chatbot-memory.ts
/**
 * Chatbot Memory Example
 *
 * This example demonstrates:
 * - Building a stateful chatbot with RecallBricks
 * - Storing conversation history
 * - Retrieving relevant context
 * - Using predictive recall for smart responses
 * - Session management
 */

import { RecallBricks } from 'recallbricks';

const rb = new RecallBricks(process.env.RECALLBRICKS_API_KEY!);

// Simulated user sessions
const userId = 'user_12345';
const sessionId = `session_${Date.now()}`;

async function main() {
  try {
    console.log('💬 RecallBricks Chatbot Memory Example\n');
    console.log(`User: ${userId}`);
    console.log(`Session: ${sessionId}\n`);

    // ============================================
    // Initialize: Store user preferences
    // ============================================
    console.log('=== Session Start ===\n');

    await rb.memories.createBatch([
      {
        content: 'User prefers concise responses',
        metadata: {
          user_id: userId,
          type: 'preference',
          category: 'communication_style'
        }
      },
      {
        content: 'User is a senior software engineer',
        metadata: {
          user_id: userId,
          type: 'profile',
          category: 'professional_info'
        }
      },
      {
        content: 'User timezone: PST (UTC-8)',
        metadata: {
          user_id: userId,
          type: 'profile',
          category: 'location'
        }
      }
    ]);

    console.log('✓ User preferences initialized\n');

    // ============================================
    // TURN 1: First message
    // ============================================
    console.log('--- Turn 1 ---\n');
    const turn1 = await handleMessage('What are the best practices for API design?', 1);
    console.log();

    // ============================================
    // TURN 2: Follow-up question
    // ============================================
    console.log('--- Turn 2 ---\n');
    const turn2 = await handleMessage('How about authentication specifically?', 2);
    console.log();

    // ============================================
    // TURN 3: Reference to earlier context
    // ============================================
    console.log('--- Turn 3 ---\n');
    const turn3 = await handleMessage('Can you give me a code example?', 3);
    console.log();

    // ============================================
    // Session End: Summarize conversation
    // ============================================
    console.log('=== Session End ===\n');

    const conversationMemories = await rb.memories.search({
      query: 'conversation',
      metadata: {
        user_id: userId,
        session_id: sessionId
      },
      weights: { semantic: 0.3, recency: 0.7 },
      limit: 10
    });

    console.log('Conversation Summary:');
    console.log(`  Total turns: 3`);
    console.log(`  Topics discussed: API design, authentication, code examples`);
    console.log(`  Memories stored: ${conversationMemories.length}\n`);

    // ============================================
    // New Session: Returning user
    // ============================================
    console.log('=== New Session (User Returns) ===\n');

    const newSessionId = `session_${Date.now() + 1000}`;

    console.log(`New Session: ${newSessionId}\n`);
    console.log('User: "Hey, I\'m back!"\n');

    // Use predictive recall to understand what user might need
    const prediction = await rb.metacognition.predict({
      context: `
        Returning user greeting.
        User ID: ${userId}
        Previous session: Discussed API design and authentication
        Session start
      `
    });

    console.log('AI Prediction for returning user:');
    console.log(`  Confidence: ${(prediction.confidence * 100).toFixed(1)}%\n`);

    console.log('Suggested Context:');
    prediction.suggestedMemories
      .filter(m => m.confidence > 0.8)
      .slice(0, 3)
      .forEach((mem, index) => {
        console.log(`  ${index + 1}. ${mem.content}`);
        console.log(`     Confidence: ${(mem.confidence * 100).toFixed(1)}%`);
      });

    console.log('\nBot Response:');
    console.log('  "Welcome back! Last time we discussed API design and authentication.');
    console.log('   Would you like to continue with code examples?"');
    console.log();

    // ============================================
    // Summary
    // ============================================
    console.log('✅ Chatbot Memory Example Completed!\n');
    console.log('Key features demonstrated:');
    console.log('  • Store conversation history automatically');
    console.log('  • Retrieve relevant context using semantic search');
    console.log('  • Use predictive recall for smart suggestions');
    console.log('  • Maintain user preferences across sessions');
    console.log('  • Track conversation flow with metadata');
    console.log('  • Welcome returning users with context awareness');

  } catch (error: any) {
    console.error('❌ Error:', error.message);
    process.exit(1);
  }
}

/**
 * Handle a single chatbot turn
 */
async function handleMessage(userMessage: string, turnNumber: number): Promise<void> {
  console.log(`User: "${userMessage}"`);

  // 1. Use predictive recall to get relevant context
  const prediction = await rb.metacognition.predict({
    context: `
      User message: "${userMessage}"
      Turn: ${turnNumber}
      User ID: ${userId}
      Session ID: ${sessionId}
    `,
    limit: 3,
    minConfidence: 0.7
  });

  console.log('\nRelevant Context Retrieved:');
  prediction.suggestedMemories.forEach((mem, index) => {
    console.log(`  ${index + 1}. ${mem.content}`);
  });

  // 2. Generate response (simulated)
  const botResponse = generateResponse(userMessage, prediction.suggestedMemories);

  console.log(`\nBot: "${botResponse}"`);

  // 3. Store this exchange as memory
  await rb.memories.createBatch([
    {
      content: `User asked: "${userMessage}"`,
      metadata: {
        user_id: userId,
        session_id: sessionId,
        turn: turnNumber,
        type: 'user_message',
        timestamp: new Date().toISOString()
      }
    },
    {
      content: `Bot responded: "${botResponse}"`,
      metadata: {
        user_id: userId,
        session_id: sessionId,
        turn: turnNumber,
        type: 'bot_message',
        timestamp: new Date().toISOString()
      }
    }
  ]);

  console.log('\n✓ Conversation stored in memory');
}

/**
 * Simulate response generation
 */
function generateResponse(message: string, context: any[]): string {
  // In a real chatbot, this would use an LLM with the retrieved context
  if (message.toLowerCase().includes('api design')) {
    return 'Best practices for API design include: RESTful principles, versioning, clear documentation, consistent naming, and proper error handling.';
  } else if (message.toLowerCase().includes('authentication')) {
    return 'For authentication, I recommend using OAuth 2.0 or JWT tokens. Bearer tokens in the Authorization header are standard practice.';
  } else if (message.toLowerCase().includes('code example')) {
    return 'Here\'s a quick example: Authorization: Bearer your_token_here';
  } else {
    return 'I can help you with that! Let me provide some information based on your preferences for concise responses.';
  }
}

main();

/**
 * To run this example:
 *
 * 1. Install dependencies:
 *    npm install recallbricks dotenv
 *
 * 2. Set your API key:
 *    export RECALLBRICKS_API_KEY='rb_live_your_key_here'
 *
 * 3. Run the script:
 *    npx ts-node chatbot-memory.ts
 *
 * This demonstrates:
 *   - How to build a stateful chatbot
 *   - Automatic context retrieval
 *   - Cross-session memory
 *   - Personalized responses based on preferences
 *   - Smart predictions for returning users
 */
